gnucash/register/register-gnome/completioncell-gnome.c

```diff
@@ -0,0 +1,1019 @@
+    g_return_if_fail (markup);
+
+    gtk_list_store_append (store, &iter);
+
+    gtk_list_store_set (store, &iter, TEXT_COL, string,
+                                      TEXT_MARKUP_COL, markup,
+                                      WEIGHT_COL, weight, -1);
+}
+
+static char*
+normalize_and_fold (char* utf8_string)
+{
+    char *normalized = g_utf8_normalize (utf8_string, -1, G_NORMALIZE_NFC);
+    if (!normalized)
+        return NULL;
+
+    char *folded = g_utf8_casefold (normalized, -1);
+    g_free (normalized);
+    return folded;
+}
+
+static void
+add_item (gpointer key, gpointer value, gpointer user_data)
+{
+    PopBox* box = user_data;
+    gchar *hash_entry = g_strdup (key);
+
+    if (hash_entry && *hash_entry)
+    {
+        gnc_utf8_strip_invalid_and_controls (hash_entry);
+        gchar *key_norm_fold = normalize_and_fold (hash_entry);
+        gchar *found = g_strstr_len (key_norm_fold, -1, box->newval);
+
+        if (found)
+        {
+            gchar *markup = NULL, *prefix = NULL, *match = NULL, *suffix = NULL;
+            gint prefix_len;
+            gint occurrence_difference;
+            gint weight;
+            gboolean have_boundary = FALSE;
+            glong len = g_utf8_strlen (box->newval, -1);
+            gulong loc = g_utf8_pointer_to_offset (key_norm_fold, found);
+            OccurrenceData *od = (OccurrenceData*)value;
+
+            if (box->register_is_reversed)
+                occurrence_difference = od->occurrence + 1;
+            else
+                occurrence_difference = box->occurrence - od->occurrence;
+
+            if (loc > 0)
+                prefix = g_utf8_substring (hash_entry, 0, loc);
+            else
+                prefix = g_strdup ("");
+            prefix_len = g_utf8_strlen (prefix, -1);
+
+            match = g_utf8_substring (hash_entry, loc, loc + len);
+
+            if (loc > 1)
+            {
+                gunichar prev = g_utf8_get_char (g_utf8_offset_to_pointer (hash_entry, loc - 1));
+                if (prev && (g_unichar_isspace (prev) || g_unichar_ispunct (prev)))
+                    have_boundary = TRUE;
+            }
+
+            suffix = g_utf8_substring (hash_entry, loc + len, g_utf8_strlen (hash_entry, -1));
+
+            if (g_utf8_strlen (suffix, -1) == 0)
+                markup = g_markup_printf_escaped ("%s<b>%s</b>%s", prefix, match, " ");
+            else
+                markup = g_markup_printf_escaped ("%s<b>%s</b>%s%s", prefix, match, suffix, " ");
+
+            if (((prefix_len != 0 ) && !have_boundary) || (box->time_now - od->date > 157680000)) // aprox 5 years
+                weight = SEPARATOR + occurrence_difference;
+            else
+                weight = occurrence_difference; // sorted by recent first
+
+            if (g_strcmp0 (key_norm_fold, box->newval) == 0) // exact match
+                weight = 1;
+
+            if (weight > SEPARATOR)
+                box->add_separator = TRUE;
```

Comment:
What is the added benefit of adding a separator between entries that are newer than 5 years and those that are older and/or where the match isn't on a word boundary ?

Personally I agree with @jralls we should just limit the list of presented suggestions. One way to achieve that would be to
*  add a visible column in the treemodel
* default-init this column to false here
* later after all entries are added,set the first x rows' visible column to true
* use a TreeModelFilter adaptor so you treeview only displays the first x columns

A reasonable, though equally arbitrary maximum number for suggestions could be 10. So you'd limit the number of visible rows to 12 (which is 10 suggestions, the `Don't autocomplete` entry and the separator).

Reply:
A faster way to achieve that: Stop stuffing the tree model after n entries.

But @Bob-IT says that he wants to see the older and middle-of-the-word matches.

Reply:
I added the separator to give a visual indication that the sort order is different in the two sections.
In the top it is the entries that have the keyed text at the start of a word boundary and sorted in most recent order.
In the bottom it is all other entries sorted in most recent order and yes entries over the time limit are also there.

If you only want the top section then there is no point in adding the rest of the entries to the list store in the first place.
I also think if we restrict what is being shown in the list we will just be replicating what we have now, trying to guess what that old description entry should be and end up creating new one.

Reply:
> I also think if we restrict what is being shown in the list we will just be replicating what we have now, trying to guess what that old description entry should be and end up creating new one.

That's solved by not setting an arbitrary time limit  and restricting the count. Once you've typed enough to get the list small enough the old entries will appear.

> I forgot to add, the more you type, the list gets smaller. With my live data file, generally after typing one character the correct entry is selected or near the top, by three characters the list is down to one or two. As I mentioned before, I normally enter a three character string, so for me it would just be 'gar' tab for my 'Garage' transaction description autocomplete.

Making the point. ;-)

More broadly let's keep in mind that this is intended to speed up data entry, not a Find substitute. If there's an old long description that you're struggling to remember do a regex search in Find. Same if you have a bunch of almost-the-same descriptions that you want to rationalize: You can fix the descriptions in a Find results register, you can't in a quickfill popup.

Reply:
Oh well, I will remove the separator and limit the list to matches for start of word boundaries.